src/read.jl

```diff
@@ -105,6 +105,6 @@ load(x) = raise_recursive(parse(x))
 
 function roundtrip(x)
   buf = IOBuffer()
-  bson(buf, Dict(:data => x))
+  bson(buf, BSONDict(:data => x))
```

Comment:
Once the `lower` was fixed, this caused another test failure.

Reply:
Shouldn't
```julia
bson(buf, Dict(:data => x))
```
not working be considered a bug?

Reply:
The failure it caused was:
https://github.com/JuliaIO/BSON.jl/blob/7f15a4b34517781622f24ce459ef89f583ed9213/test/runtests.jl#L54
(it errored).

But note that `bson(buf, Dict(:data=>x))` does work.

Reply:
> But note that bson(buf, Dict(:data=>x)) does work.

but it wouldn't work for `x = T(())`. I'm not sure about the implications of that but it doesn't seem safe to break this right?

Reply:
Yep, you're right.  Here the example:
```julia
julia> using BSON

julia> struct T{A}
         x::A
       end

julia> x = T(())
T{Tuple{}}(())

julia> buf = IOBuffer()
IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1)

julia> bson(buf, Dict(:data => x))

julia> BSON.load(seek(buf, 0))[:data]
ERROR: MethodError: no method matching T{Tuple{}}()
Closest candidates are:
  T{Tuple{}}(::Any) where A at REPL[2]:2
```
Well, you seem to have a better grasp of this than me, maybe you can take this (or something else) and fix it?

Reply:
I don't have any solutions, I'm not familiar with the internals. Let's leave this PR open hoping someone comes in with the correct solution at some point

Reply:
Then the question is whether this "bug" is bigger than the bug #26.  Otherwise merging this might still make sense.